### DISCUSSION TITLE
How client and server negotiates OAEP padding.?

### DISCUSSION THREAD

**Discussion by suresh043** [openssl/openssl#26104] [Q&A] [answered]
Hi Experts,

I am trying to find out answers for below questions. 
Point under consideration:
 - RSA key-exchange.
 - TLS 1.2 handshake between geographically separated client and server.
 
Say, I have client A which use RSA-PKCS#1-v1.5 padding for PMS encryption with public key of server X.
And client B, which uses RSAES-OAEP padding for PMS encryption with public key of server X.

Now, when request comes to the server for PMS decryption how server knows which padding is used by the client.
Since, RSA_private_decrypt() API taking padding as a argument beforehand only. 

I know RSA-PKCS#1 v1.5 has formatted. But that also we will come to know after decryption only.

//CLIENT
MSG: Hi
PM : 0x00 | 0x02 | non-zero random bytes | 0x00 | "Hi <ascii>""     // RSAPKCS#1 v1.5 PM (padded message)
CT = encryption(PM)

//SERVER 
PM = decryption(CT)
unpad message and get MSG (Hi)

Follow up questions:
1) Is it possible to have RSA-PKCS#1 v1.5 and OAEP side-by-side.
2) Does OpenSSL's SSL stack (libssl) supports RSA-OAEP ? If not is it due to point-1 is not possible.

**Comment by suresh043** [comment]
cc: @t8m

**Comment by t8m** [comment, marked as answer]
The RSA OAEP padding cannot be used in a handshake with TLS-1.2 RSA key exchange, because the TLS-1.2 RFC specifies only the PKCS#1 v1.5 padding there.

**Comment by suresh043** [comment]
Thank you.
